### DISCUSSION TITLE
Gradient Operator Dotted with Normal Vector

### DISCUSSION THREAD

**Discussion by TLWise** [idaholab/moose#28566] [Q&A Modules: General]
This is a different spin to a previous question to satisfy my need to code to the PDE formulation for the problem directly in MOOSE. Given the equation (I decided to derive it in PDE form) represented in the picture, is there a already defined post processor that will provide this solution?

![image](https://github.com/user-attachments/assets/e7eabc70-cc4d-4071-b63f-3c6c11162f89)

**Comment by GiudGiud** [comment]
Hello

I dont know about one. I dont think this would be very difficult to implement. You should try and report on your progress and difficulties here

**Reply by TLWise** [reply]
Thank you for the vote of confidence that a solution should be easily at hand. I'll keep the group informed of my challenges and victories.

**Reply by TLWise** [reply]
It appears that VectorNeumannBC.h has elements of what I need to accomplish solving this system of equations. It appears that my attempting to use the VectorNeumannBC.h code as a template for a Postprocessor causes the system to crash. How can I get the necessary mathematical elements from VectorNeumannBC.h, without confusing the BCs vs PP during the MOOSE processing?

**Reply by TLWise** [reply]
Here are the current errors associated with my current efforts:
```
/home/twise/mambaforge3/envs/moose/x86_64-conda-linux-gnu/include/c++/10.4.0/ext/new_allocator.h:156:4: error: invalid new-expression of abstract class type 'VectorPartialNormal_crab'
  156 |  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/twise/projects/crab/src/postprocessors/VectorPartialNormal_crab.C:10,
                 from /home/twise/projects/crab/build/unity_src/postprocessors_Unity.C:17:
/home/twise/projects/crab/build/header_symlinks/VectorPartialNormal_crab.h:21:7: note:   because the following virtual functions are pure within 'VectorPartialNormal_crab':
   21 | class VectorPartialNormal_crab : public ADInterfaceKernel
      |       ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/twise/projects/crab/build/header_symlinks/VectorPartialNormal_crab.h:14,
                 from /home/twise/projects/crab/src/postprocessors/VectorPartialNormal_crab.C:10,
                 from /home/twise/projects/crab/build/unity_src/postprocessors_Unity.C:17:
/home/twise/projects/moose/framework/build/header_symlinks/ADInterfaceKernel.h:68:18: note:     'ADReal ADInterfaceKernelTempl<T>::computeQpResidual(Moose::DGResidualType) [with T = double; ADReal = MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<64> >, true>]'
   68 |   virtual ADReal computeQpResidual(Moose::DGResidualType type) = 0;
      |                  ^~~~~~~~~~~~~~~~~
make: *** [/home/twise/projects/moose/framework/build.mk:151: /home/twise/projects/crab/build/unity_src/postprocessors_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
```

**Reply by GiudGiud** [reply]
>    21 | class VectorPartialNormal_crab : public ADInterfaceKernel

this is not an appropriate base class for a postprocessor.
Are you trying to implement a kernel for this (for creating an equation)? Or a postprocessor (for output only)

Is the boundary this object will be active on in the middle of the mesh, in between two domains, both that you use information from?
I dont think so right? It s more of a boundary thing, with information from a single domain?

The error is telling you that the class you implemented is "virtual" so it cannot be registered. It is virtual because you did not implement this routine `   virtual ADReal computeQpResidual(Moose::DGResidualType type);`. 
It's ok because I dont think you should inherit from interface kernels

**Reply by TLWise** [reply]
Is there a MOOSE object that will allow me to calculate the gradient of (v^2+w^2)^1/2 on a boundary, where v and w are components of vector V?

**Reply by GiudGiud** [reply]
No this is very specific. 

Do you want to see a field or an integral / a single number?

**Reply by TLWise** [reply]
I would like to calculate the surface integral of this formulation and graph it. I have Aux Variables that represent each component.

**Reply by GiudGiud** [reply]
So graph it through time.

Do you mean to plot the value of each compoent?
Or gradient times the normal of the surface?

**Reply by TLWise** [reply]
Yes, plot the values of the integral of the gradient of the function time the normal of the surface. Preferably in a PP.

**Reply by GiudGiud** [reply]
I think the SideDiffusiveFluxIntegral gets you almost there.

However you ll need to define the variable to take the gradient of in a new variable (using a ParsedMaterial for example) and it should be high order enough to have a defined gradient

**Reply by TLWise** [reply]
Thank you. I will try that and report back on my progress.

**Reply by GiudGiud** [reply]
Alternatively we could do something with functors. 

But we are missing a SideVectorDiffusiveFluxIntegral object. If you code it we can add it to moose.
We would first form each component of the gradient using 3 ParsedFunctorMaterial, then use a GenericVectorFunctorMaterial to assemble them.

**Reply by TLWise** [reply]
I am attempting to develop a vector based Postprocessor using the following code that inherits from SideIntegralVariablePostprocessor:

''' template <bool is_ad, typename T>
Real
SideDiffusiveFluxIntegral_crabTempl<is_ad, T>::computeQpIntegral()
{
    using MetaPhysicL::raw_value;

    const auto state = determineState();
    const Moose::ElemSideQpArg side_arg = {_current_elem, _current_side, _qp, _qrule, _q_point[_qp]};
    const auto vel_x = raw_value(_vel_x(side_arg, state));
    const auto vel_y = _vel_y ? raw_value((*_vel_y)(side_arg, state)) : 0;
    const auto vel_z = _vel_z ? raw_value((*_vel_z)(side_arg, state)) : 0;

    if (_advected_variable_supplied)
    {

        auto & variable = getCoupledMooseVars();
        if (std::any_of(variable.begin(), variable.end(), [](auto & var) { return !var->isNodal(); }))
            mooseError("Trying to use a non-nodal variable 'advected_variable' for side advection "
                       "integral calculation, which is currently not supported.");

        return (_use_normal
                ? _advected_variable[_qp] * RealVectorValue(vel_x, vel_y, vel_z) * _normals[_qp]
                : _advected_variable[_qp] * RealVectorValue(vel_x, vel_y, vel_z)(_component));
    }
    else if (_diffusion_coef_supplied)
        return (_use_normal ? raw_value(_diffusion_coef[_qp]) *
                              RealVectorValue(vel_x, vel_y, vel_z) * _normals[_qp]
                            : raw_value(_diffusion_coef[_qp]) *
                              RealVectorValue(vel_x, vel_y, vel_z)(_component));
    else
        return (_use_normal ? RealVectorValue(vel_x, vel_y, vel_z) * _normals[_qp]
                            : RealVectorValue(vel_x, vel_y, vel_z)(_component));
} '''

I am getting the following error:
''' error: no match for 'operator*' (operand types are 'MetaPhysicL::RawType<MaterialProperty<double>, void>::value_type' {aka 'MaterialProperty<double>'} and 'libMesh::RealVectorValue' {aka 'libMesh::VectorValue<double>'})
   93 |         return (_use_normal ? raw_value(_diffusion_coef[_qp]) *
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
   94 |                               RealVectorValue(vel_x, vel_y, vel_z) * _normals[_qp]
 '''

**Reply by GiudGiud** [reply]
Can you paste the header as well ?

The type of the property must not be consistent with using the multiplication with *

**Reply by TLWise** [reply]
Sure thank you. Here it is.:

''' 
#pragma once

// MOOSE includes
#include "SideIntegralVariablePostprocessor.h"

// Forward Declarations
template <bool, typename>
class SideDiffusiveFluxIntegral_crabTempl;
typedef SideDiffusiveFluxIntegral_crabTempl<false, Real> SideDiffusiveFluxIntegral_crab;
typedef SideDiffusiveFluxIntegral_crabTempl<true, Real> ADSideDiffusiveFluxIntegral_crab;
typedef SideDiffusiveFluxIntegral_crabTempl<false, RealVectorValue> SideVectorDiffusivityFluxIntegral;
typedef SideDiffusiveFluxIntegral_crabTempl<true, RealVectorValue> ADSideVectorDiffusivityFluxIntegral;

/**
 * This postprocessor computes a side integral of the mass flux.
 */
template <bool is_ad, typename T>
class SideDiffusiveFluxIntegral_crabTempl : public SideIntegralVariablePostprocessor
{
public:
    static InputParameters validParams();

    SideDiffusiveFluxIntegral_crabTempl(const InputParameters & parameters);

protected:
    Real computeQpIntegral() override;

    const GenericMaterialProperty<T, is_ad> * const _diffusion_coef;
    /// Whether the normal component has been selected
    const bool _use_normal;

    /// Will hold 0, 1, or 2 corresponding to x, y, or z.
    const int _component;
    /// Whether an advected variable was supplied in the input
    const bool _advected_variable_supplied;
    /// Whether a diffusion coefficient variable was supplied in the input
    const bool _diffusion_coef_supplied;
    /// Variable storing the advected quantity; used for finite elements
    const VariableValue & _advected_variable;

    /// Velocity components
    const Moose::Functor<Real> & _vel_x;
    const Moose::Functor<Real> * const _vel_y;
    const Moose::Functor<Real> * const _vel_z;

//private:
//    /// Routine to get the diffusive flux with a Real diffusivity
//    RealVectorValue diffusivityGradientProduct(const RealVectorValue & grad_u, Real diffusivity);
//
//    /// Routine to get the diffusive flux with a RealVectorValue diffusivity
//    RealVectorValue diffusivityGradientProduct(const RealVectorValue & grad_u,
//                                               const RealVectorValue & diffusivity);
}; ''

**Reply by GiudGiud** [reply]
_diffusion_coef is a pointer so you need to do
```
(*_diffusion_coef)[_qp])
```
to dereference it

**Reply by TLWise** [reply]
Here is the new error: 

''' error: cannot convert 'libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<double> >::type' {aka 'libMesh::TypeVector<double>'} to 'libMesh::Real' {aka 'double'} in return
   93 |         return (_use_normal ? raw_value((*_diffusion_coef)[_qp]) *
      |                ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                             |
      |                             libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<double> >::type {aka libMesh::TypeVector<double>}
   94 |                               RealVectorValue(vel_x, vel_y, vel_z) * _normals[_qp]
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   95 |                             : raw_value((*_diffusion_coef)[_qp]) *
      |                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   96 |                               RealVectorValue(vel_x, vel_y, vel_z)(_component)); '''

**Reply by GiudGiud** [reply]
right now you have 3 vectors multiplying each other, which does not end up with a single value like you want

D is a vector
normal is vector
RealVectorValue(vel_x, vel_y, vel_z) is a vector

I think you need to take the component of one these or something

**Comment by TLWise** [comment]
Got it. I will remove one to work towards a solution. Is it possible to feed the result into a vector that stores the result? I still need to apply grad_u to rhe velocity vector?

**Reply by GiudGiud** [reply]
You can store the result with :
auto result = ....

then modify it then use
```
return a scalar;
```
to set up the output of the routine

**Reply by TLWise** [reply]
Given the following code:

```
 template <bool is_ad>
Real
SideDiffusiveFluxIntegral_crabTempl<is_ad>::computeQpIntegral()

//{
//    return _grad_u[_qp];
//}
{
    using MetaPhysicL::raw_value;

    const auto state = determineState();
    const Moose::ElemSideQpArg side_arg = {_current_elem, _current_side, _qp, _qrule, _q_point[_qp]};
    const auto vel_x = raw_value(_vel_x(side_arg, state));
    const auto vel_y = _vel_y ? raw_value((*_vel_y)(side_arg, state)) : 0;
    const auto vel_z = _vel_z ? raw_value((*_vel_z)(side_arg, state)) : 0;
    const auto vel_t = sqrt((vel_x*vel_x)+(vel_y*vel_y)+(vel_z*vel_z));

    if (_advected_variable_supplied)
    {

        auto & variable = getCoupledMooseVars();
        if (std::any_of(variable.begin(), variable.end(), [](auto & var) { return !var->isNodal(); }))
            mooseError("Trying to use a non-nodal variable 'advected_variable' for side advection "
                       "integral calculation, which is currently not supported.");

        return (_use_normal
                ? _advected_variable[_qp] * RealVectorValue(vel_x, vel_y, vel_z) * _normals[_qp]
                : _advected_variable[_qp] * RealVectorValue(vel_x, vel_y, vel_z)(_component));
    }
    else if (_advected_mat_prop_supplied)
        return (_use_normal ? raw_value(_advected_material_property[_qp]) *
                              RealVectorValue(vel_x, vel_y, vel_z) * _normals[_qp]
                            : raw_value(_advected_material_property[_qp]) *
                              RealVectorValue(vel_x, vel_y, vel_z)(_component));
    else
        return (_use_normal ? RealVectorValue(vel_x, vel_y, vel_z) * _normals[_qp]
                            : RealVectorValue(vel_x, vel_y, vel_z)(_component));
}
```

Is it possible to determine the gradient of the RealVectorValue(vel_x, vel_y, vel_z) or maybe determine the gradient of each component then combine them into a RealVectorValue(grad(vel_x), grad(vel_y), grad(vel_z)) ?

**Reply by GiudGiud** [reply]
No to the second one. The gradient of a RealVectorValue is a RealTensorValue not a RealVectorValue of Gradients

To get the gradient of vel_x you can do:
```
raw_value(_vel_x.gradient(side_arg, state))
```
and vel_y you can do:
```
raw_value(_vel_y->gradient(side_arg, state))
```
Side note: you ll want to use back quotes ` not straight quotes ' for code snippets

**Reply by TLWise** [reply]
For this instance of the code `raw_value(_vel_x.gradient(side_arg, state))` does it represent, d/dx(vel_x)i+d/dy(vel_x)j+d/dz(vel_x)k?

**Reply by TLWise** [reply]
Is the operation allowed to take a RealVectorValue, C and get the gradient of it with a operation similar to  `C.gradient(side_arg, state))` ?

**Reply by GiudGiud** [reply]
For your first message, no the raw value of a vector is still a vector.

For your second message, no. The .gradient(location, state) is only available for variables

**Reply by TLWise** [reply]
For the first instance are the components of the vector composed of [ d/dx(vel_x), d/dy(vel_x), d/dz(vel_x)]?

**Reply by GiudGiud** [reply]
Yes

**Reply by TLWise** [reply]
Got it. Thank your for your patience and continued mentoring as I attempt to develop MOOSE FE results for these types of problems.

**Reply by TLWise** [reply]
In developing my modeling method for determining the shear stress at a boundary, I have defaulted to using the problem of pressure driven flow between fixed plates. I have run a comparison study using the MOOSE FV method and FEM method, using the same material parameters, geometry, solvers etc. The shear stress calculated by the FV method is in very good agreement with the analytical value, my FEM calculation has about a 1.33% error. What stands out most to me is that the max pressure in the FV model is almost twice that of the FEM solution. Is this level of error expected between the two methods regarding pressure?

**Reply by GiudGiud** [reply]
> What stands out most to me is that the max pressure in the FV model is almost twice that of the FEM solution.

Is the pressure drop correct though? As in is this max pressure just a local maxima?
Is the mesh converged? And are both solutions not showing any oscillation?

**Reply by TLWise** [reply]
Both solutions demonstrate no oscillation. The pressure drop is established by an analytical inlet velocity function and a zero pressure outlet boundary conditions.
![Screenshot from 2024-11-22 05-11-50](https://github.com/user-attachments/assets/dcc2397a-928f-4157-b3bf-7529e8e1d023)
![Screenshot from 2024-11-22 05-14-53](https://github.com/user-attachments/assets/6c8f3914-6eb4-49d8-8611-3e19fa5c72ff)

![Screenshot from 2024-11-22 05-15-26](https://github.com/user-attachments/assets/fb2073a7-541d-416d-a217-d7d2248c2962)

**Reply by GiudGiud** [reply]
And you have twice the pressure drop in FV. 
the definition of the friction factors are different in both discretizations
Are you using our friction kernels ?

or is the pressure drop only from the no slip on the channel walls?

**Reply by GiudGiud** [reply]
can you please share both input files? maybe we can spot what's wrong

**Reply by TLWise** [reply]
The pressure drop is only from the no slip boundary condition. I will send you the files needed for each.

**Reply by TLWise** [reply]
Was there ever a determination on what made the solution for these two method different, based on the code provided?

**Reply by GiudGiud** [reply]
Hello

Where did you send the code? / the files?

**Comment by TLWise** [comment]
Thank you for taking a look at this for me. Let me know if I left something out that prevents you from running the code.

T. Leon Wise






________________________________

**Reply by TLWise** [reply]
@GiudGiud,  I recently had a similar result with another benchmark where the pressure of my model resulted in a pressure that was approximately double that of the benchmark. It appears that the difference was related to prescribing the pressure at the outlet as equal to zero vs a boundary condition of do nothing relative to the velocity at the outlet.

**Reply by GiudGiud** [reply]
Are you imposing the outlet velocity and the outlet pressure?

When not imposing the outlet pressure, do you set the pressure anywhere?
For incompressible flow, the pressure can be shifted by any constant and still be a solution, so we need the pressure to be set somewhere

**Reply by TLWise** [reply]
Yes. I impose a zero pressure condition at a single outlet node.

**Comment by TLWise** [comment]
***@***.***


Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by GiudGiud** [reply]
I think git might have obscured that address
